facilitator/src/gcp_oauth.rs

```diff
@@ -84,17 +91,231 @@ struct ServiceAccountKeyFile {
     token_uri: String,
 }
 
+/// This enum captures the different ways that a GcpOauthTokenProvider can
+/// obtain the default Oauth token, used either to authenticate to GCP services
+/// or to obtain a further service account Oauth token from GCP IAM.
+enum DefaultTokenProvider<P: ProvideAwsCredentials> {
+    /// Fetch a token from the GKE metadata service for the GCP SA mapped to
+    /// the current Kubernetes SA via GKE workload identity (not to be confused)
+    /// with workload identity *pool*). Used by workloads in Google Kubernetes
+    /// Engine.
+    GkeMetadataService,
+    /// Use a GCP service account key file to authenticate to GCP IAM as some
+    /// service account.
+    ServiceAccountKeyFile(ServiceAccountKeyFile),
+    /// Use a GCP Workload Identity Pool to federate GCP IAM with AWS IAM,
+    /// allowing workloads in AWS Elastic Kubernetes Service to impersonate a
+    /// GCP SA.
+    AwsIamFederationViaWorkloadIdentityPool(WorkloadIdentityPoolParameters<P>),
+}
+
+impl<P: ProvideAwsCredentials> fmt::Debug for DefaultTokenProvider<P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut debug_struct = f.debug_struct("DefaultTokenProvider");
+        match self {
+            Self::GkeMetadataService => debug_struct.field(
+                "token_source",
+                &"default service account from GKE metadata service".to_owned(),
```

Comment:
This is an interesting construction. It looks like this param wants `&dyn Debug`. But `str` has a Debug impl: https://doc.rust-lang.org/nightly/std/primitive.str.html#impl-Debug. Any idea why a plain string literal wouldn't work here? Not a requirement for merging, just kibbitzing really.

Reply:
If I do

Reply:
This is probably not the strict way to think about it, but maybe:

`&str` is already a fat pointer. `&dyn Debug` needs to be a fat pointer. There's not enough "fat" left in `&str` to stick the vtable in, so it needs another `&` where it can stick in some fat. :-D